### PR TITLE
Add the WooPay Direct Checkout flow to the blocks mini cart widget

### PR DESCRIPTION
**Pull request by ricardo**
Fixes 2647-gh-Automattic/woopay

### Changes proposed in this Pull Request

- Enqueue the direct checkout scripts if the current page has triggered the `woocommerce_blocks_cart_enqueue_data` action, which is used by the blocks mini cart and cart as widgets ([src](https://github.com/woocommerce/woocommerce/blob/94ed6a943fd593b58d1ffc9005784ee164f30f0e/plugins/woocommerce-blocks/docs/third-party-developers/extensibility/hooks/actions.md?plain=1#L241-L254)).
- Handle the opening of the mini cart through JavaScript and add the direct checkout event listeners to the "Go to checkout" button.

**Note:** In this PR we are not adding support for the legacy mini cart `woocommerce_mini_cart()` yet.

### Testing instructions

**Setup**
1. Ensure the WooPay direct checkout feature flag `_wcpay_feature_woopay_direct_checkout` is set to `1`.
2. Switch to this branch and run `npm run build:client`.
3. As a merchant, navigate to **wp-admin > Appearance > Widgets**.
4. Add the "Mini-cart" widget to any available section and hit "Update".

**Test 1: Ensure the direct checkout scripts are enqueued in the right pages**
1. As a shopper, add a product to the cart.
2. Navigate to any page where the Mini-cart widget is available.
3. Ensure the `WCPAY_WOOPAY_DIRECT_CHECKOUT` script is enqueued in the page source.
4. As a merchant, remove the Mini-cart widget temporarily.
5. Navigate to any page besides the cart and ensure the `WCPAY_WOOPAY_DIRECT_CHECKOUT` script is NOT enqueued in the page source.

**Test 2: Ensure the direct checkout flow works in the mini cart**
1. As a shopper, add a product to the cart.
2. Navigate to any page where the mini cart is visible (any page besides the cart or the checkout).
3. Click the mini cart "Go to checkout" button.
4. Ensure the direct checkout flow works as expected:
  - If the user is logged in to WooPay, the user is redirected to WooPay.
  - If the user is logged out from WooPay, the user is redirected to the merchant checkout.
  - Please test with third party cookies enabled vs. disabled.

**Test 3: Ensure the direct checkout flow works from classic and blocks checkout**
1. As a shopper, add a product to the cart.
2. Ensure the direct checkout flow works from classic and blocks carts:
  - If the user is logged in to WooPay, the user is redirected to WooPay.
  - If the user is logged out from WooPay, the user is redirected to the merchant checkout.
  - Please test with third party cookies enabled vs. disabled.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8795` or branch name `add/woopay-direct-checkout-to-minicart` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=add/woopay-direct-checkout-to-minicart) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 5ce2d84f3c3ee2514995758f8f1ff6e75451a7fe
- Build time: 2024-05-15 02:22:55 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: client/checkout/woopay/direct-checkout/index.js [hunk 1]
```diff
@@ -9,59 +9,116 @@ import { debounce } from 'lodash';
...
+							addMiniCartEventListener
+						);
+						return;
+					}
+				}
+			}
+		}
+	} );
+
+	observer.observe( document.body, { childList: true } );
```
Comment (ricardo) [inline client/checkout/woopay/direct-checkout/index.js:98]: Alternatively, we could add a click event listener to the mini cart button, but that would fail if the button gets re-rendered, so I figured the most reliable way would be through a `MutationObserver`.

There can be some performance overhead to attach the observer to `document.body` in case the website has too many DOM changes constantly going on for some reason, but we're doing our best here by passing only `childList` as `true`, and not `subtree` or `attributes` types of events.

So basically, we're only iterating nodes that are being added or removed from `document.body` and that are immediate children.

File: client/checkout/woopay/direct-checkout/index.js [hunk 2]
```diff
@@ -9,59 +9,116 @@ import { debounce } from 'lodash';
...
+	maybeObserveMiniCart();
+
+	const checkoutButtons = WooPayDirectCheckout.getCheckoutButtonElements();
+	handleWooPayDirectCheckout( checkoutButtons );
 } );
 
 jQuery( ( $ ) => {
 	$( document.body ).on( 'updated_cart_totals', async () => {
-		if (
-			! WooPayDirectCheckout.isWooPayDirectCheckoutEnabled() ||
```
Comment (ricardo) [inline client/checkout/woopay/direct-checkout/index.js:46]: Context for this change: p1715106668797959-slack-C01BZUL57SQ

File: includes/class-wc-payments-woopay-direct-checkout.php [hunk 3]
```diff
@@ -20,7 +20,7 @@ class WC_Payments_WooPay_Direct_Checkout {
 	 * @return void
 	 */
 	public function init() {
-		add_action( 'wp_enqueue_scripts', [ $this, 'scripts' ] );
+		add_action( 'wp_footer', [ $this, 'scripts' ] );
```
Comment (ricardo) [inline includes/class-wc-payments-woopay-direct-checkout.php:23]: This is because the mini cart widget can be added to as late as in the footer section of the template.

Comment (lovo-h) [reply includes/class-wc-payments-woopay-direct-checkout.php:23]: Nit: In order to move the script to the footer, should we change the [wp_enqueue_script()](https://developer.wordpress.org/reference/functions/wp_enqueue_script) call inside the `scripts()` function instead of this hook?

File: client/checkout/woopay/direct-checkout/index.js [hunk 4]
```diff
@@ -9,59 +9,119 @@ import { debounce } from 'lodash';
...
+/**
+ * If the mini cart widget is available on the page, observe when the drawer element gets added to the DOM.
+ *
+ * As of today, no window events are triggered when the mini cart is opened or closed,
+ * nor there are attribute changes to the "open" button, so we have to rely on a MutationObserver
+ * attached to the `document.body`, which is where the mini cart drawer element is added.
+ */
+const maybeObserveMiniCart = () => {
+	// Check if the widget is available on the page.
+	if ( ! document.querySelector( '.wc-block-mini-cart' ) ) {
```
Comment (lovo-h) [inline client/checkout/woopay/direct-checkout/index.js]: Instead of querying for the `.wc-block-mini-cart` CSS class, should we query for `woocommerce/mini-cart` data-block-name HTML element attribute?

```javascript
document.querySelector('[data-block-name="woocommerce/mini-cart"]');
```

I don't believe there's a major difference but this approach may add one more level of separation from accidentally detecting a false-positive.

File: client/checkout/woopay/direct-checkout/index.js [hunk 5]
```diff
@@ -9,59 +9,119 @@ import { debounce } from 'lodash';
...
+	WooPayDirectCheckout.addRedirectToWooPayEventListener(
+		checkoutButtons,
+		false
+	);
+};
+
+/**
+ * Add an event listener to the mini cart checkout button.
+ */
+const addMiniCartEventListener = async () => {
```
Comment (lovo-h) [inline client/checkout/woopay/direct-checkout/index.js]: Nit: Are we able to remove `async` for this function, since it's not being used?

File: client/checkout/woopay/direct-checkout/index.js [hunk 6]
```diff
@@ -9,59 +9,119 @@ import { debounce } from 'lodash';
...
+						node.nodeType === 1 &&
+						node.matches(
+							'.wc-block-components-drawer__screen-overlay'
+						)
+					) {
+						// Wait until the button is rendered and add the event listener to it.
+						waitForSelector(
+							WooPayDirectCheckout.redirectElements
+								.BLOCKS_MINI_CART_PROCEED_BUTTON,
+							addMiniCartEventListener
```
Comment (lovo-h) [inline client/checkout/woopay/direct-checkout/index.js:89]: It looks like `addMiniCartEventListener()` does not accept any parameters, but `waitForSelector()` uses `addMiniCartEventListener()` as a callback that accepts one parameter. 

This approach seems intentional, but I wanted to confirm.

File: client/checkout/woopay/direct-checkout/index.js (hunk 1 above)
Comment (lovo-h) [reply client/checkout/woopay/direct-checkout/index.js:98]: Agreed. I think this is an acceptable approach. 

Although possible, it would be odd for a site to constantly add/remove elements from the `document.body`.

File: client/checkout/woopay/direct-checkout/index.js [hunk 7]
```diff
@@ -9,59 +9,119 @@ import { debounce } from 'lodash';
...
+	handleWooPayDirectCheckout( checkoutButtons );
 } );
 
 jQuery( ( $ ) => {
 	$( document.body ).on( 'updated_cart_totals', async () => {
-		if (
-			! WooPayDirectCheckout.isWooPayDirectCheckoutEnabled() ||
-			shouldSkipWooPay()
-		) {
+		if ( shouldSkipWooPay() ) {
```
Comment (lovo-h) [inline client/checkout/woopay/direct-checkout/index.js]: This is not related to this PR but am I correct in my thinking that we may want to add similar logic (early exit when `shouldSkipWooPay()`) to the following functions, below:
* `addItemCallback()`.
* `debounceSetItemQtyCallback()`.
* `removeItemCallback()`.

Otherwise, we will be prefetching encrypted session data that will not be used.

**Review by lovo-h — CHANGES_REQUESTED** [review]
The code changes look good and have test well.

There were just a few, minor details I wanted to confirm before merging.

**Testing Instructions**
* ✅ Test 1: Ensure the direct checkout scripts are enqueued in the right pages.
* ✅ Test 2: Ensure the direct checkout flow works in the mini cart.
* ✅ Test 3: Ensure the direct checkout flow works from classic and blocks checkout.
* ✅ `npm run test:js`.

File: includes/class-wc-payments-woopay-direct-checkout.php (hunk 3 above)
Comment (ricardo) [reply includes/class-wc-payments-woopay-direct-checkout.php:23]: The issue is not about moving the scripts to the footer, but it's in detecting if `did_action( 'woocommerce_blocks_cart_enqueue_data' )` had already been fired, as the mini cart can get added to a footer column for example.

We use `woocommerce_blocks_cart_enqueue_data` to tell if the blocks mini cart is present on the page.

File: client/checkout/woopay/direct-checkout/index.js (hunk 6 above)
Comment (ricardo) [reply client/checkout/woopay/direct-checkout/index.js:89]: Yes. It's intentional as I tried to keep `waitForSelector` as generic as possible in case it gets used somewhere else.

In `addMiniCartEventListener` we could go with the callback element parameter, but since we were using `WooPayDirectCheckout.getCheckoutButtonElements()` and `WooPayDirectCheckout.getClassicProceedToCheckoutButton()`, I decided to keep the same pattern with `WooPayDirectCheckout.getMiniCartProceedToCheckoutButton()`.

File: client/checkout/woopay/direct-checkout/index.js (hunk 4 above)
Comment (ricardo) [reply client/checkout/woopay/direct-checkout/index.js]: Good call. Fixed in [5df4ae5](https://github.com/Automattic/woocommerce-payments/pull/8795/commits/5df4ae561e3530aa368f7d1353f0de52cc9e6c59).

File: client/checkout/woopay/direct-checkout/index.js (hunk 5 above)
Comment (ricardo) [reply client/checkout/woopay/direct-checkout/index.js]: Fixed in [61d6e51](https://github.com/Automattic/woocommerce-payments/pull/8795/commits/61d6e5165366d0ba14c2e361bd4654ccc2dffa08).

File: client/checkout/woopay/direct-checkout/index.js (hunk 7 above)
Comment (ricardo) [reply client/checkout/woopay/direct-checkout/index.js]: I agree.

Instead of checking for `shouldSkipWooPay`, I added the hooks in the `load` window event listener and after `isWooPayThirdPartyCookiesEnabled()` in [5ce2d84](https://github.com/Automattic/woocommerce-payments/pull/8795/commits/5ce2d84f3c3ee2514995758f8f1ff6e75451a7fe). LMK what you think.

@lovo-h – I tested this by adding a `console.log` to each callback and going to the blocks cart and adding, removing and setting the quantity of an item, however I wasn't able to hit the action hook when adding a new item from the suggestions list after the cart is emptied, am I testing this wrong? What's one way to hit the `experimental__woocommerce_blocks-cart-add-item`?

File: client/checkout/woopay/direct-checkout/index.js (hunk 6 above)
Comment (lovo-h) [reply client/checkout/woopay/direct-checkout/index.js:89]: That makes sense. Thanks for confirming.

File: client/checkout/woopay/direct-checkout/index.js (hunk 7 above)
Comment (lovo-h) [reply client/checkout/woopay/direct-checkout/index.js]: > Instead of checking for `shouldSkipWooPay`, I added the hooks in the `load` window event listener and after `isWooPayThirdPartyCookiesEnabled()` in [5ce2d84](https://github.com/Automattic/woocommerce-payments/pull/8795/commits/5ce2d84f3c3ee2514995758f8f1ff6e75451a7fe). LMK what you think.

I think this approach looks cleaner. Good call.

>  I wasn't able to hit the action hook when adding a new item from the suggestions list after the cart is emptied, am I testing this wrong? What's one way to hit the `experimental__woocommerce_blocks-cart-add-item`?

Good catch.

I also wasn't able to trigger this callback either. It appears like it was removed in WooCommerce:
* [WooCommerce#43177](https://github.com/woocommerce/woocommerce/pull/43177)

If I recall correctly, this hook was added when wcBlocks was an independent plugin from WooCommerce. In wcBlocks, it looks like the [`experimental__woocommerce_blocks-cart-add-item` hook still exists](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/assets/js/atomic/blocks/product-elements/add-to-cart/shared/add-to-cart-button.tsx#L157-L160).

One solution would be to replace the `experimental__woocommerce_blocks-cart-add-item` hook with `wc-blocks_added_to_cart`:
```javascript
document.body.addEventListener(
  'wc-blocks_added_to_cart',
  addItemCallback
);
```

However, this is an unrelated issue to this PR. So, we can create a separate GH issue.

File: includes/class-wc-payments-woopay-direct-checkout.php (hunk 3 above)
Comment (lovo-h) [reply includes/class-wc-payments-woopay-direct-checkout.php:23]: Ah, I see. That makes sense 👍

**Review by lovo-h — APPROVED** [review]
Code changes look good and test well 👍 

**Testing Instructions**
* ✅ Test 1: Ensure the direct checkout scripts are enqueued in the right pages.
* ✅ Test 2: Ensure the direct checkout flow works in the mini cart.
* ✅ Test 3: Ensure the direct checkout flow works from classic and blocks checkout.
* ✅ `npm run test:js`.

File: client/checkout/woopay/direct-checkout/index.js (hunk 7 above)
Comment (ricardo) [reply client/checkout/woopay/direct-checkout/index.js]: Created an issue. Feel free to look into this if you have the availability.

https://github.com/Automattic/woocommerce-payments/issues/8831
